salvo/src/lib/benchmark/base_benchmark.py

```diff
@@ -105,40 +159,8 @@ def run_image(self, image_name: str,
     """
     return self._docker_image.run_image(image_name, run_parameters)
 
-  def pull_images(self) -> List[str]:
-    """Retrieve all images necessary for the benchmark.
-
-    Retrieve the NightHawk and Envoy images defined in the control
-    object.
-
-    Returns:
-        a List of required image names that are retrievable. If any image is
-          unavailable, we return an empty list. The intent is for the caller to
-          build the necessary images.
-
-    Raises:
-        BenchmarkError: if a requested image is unavailable.
-    """
-    retrieved_images = []
-    images = self.get_images()
-
-    for image in [
-        images.nighthawk_benchmark_image,
-        images.nighthawk_binary_image,
-        images.envoy_image
-    ]:
-      # If the image name is not defined, we will have an empty string.
-      # For unit testing we'll keep this behavior. For true usage, we
-      # should raise an exception when the benchmark class performs its
-      # validation
-      if image:
-        retrieved_image = self._docker_image.pull_image(image)
-        log.debug(f"Retrieved image: {retrieved_image} for {image}")
-        if retrieved_image is None:
-          raise BenchmarkError("Unable to retrieve image: %s" % image)
-        retrieved_images.append(retrieved_image)
-
-    return retrieved_images
+  def execute_benchmark(self) -> None:
```

Comment:
If we are expecting derived classes to implement this method, this is the location where we should have the main doc comment establishing the expected behaviors or "contract" with the child classes. This is also where we say anything specific the child classes must do in their implementations.

Reply:
Updated.  Please let me know if more changes are needed, thanks!

Reply:
Since we expect the child classes to override this method, we should outline the expected "contract". I.e. what is this method supposed to do regardless of the implementation (can we say anything more specific than "it is supposed to execute the benchmark"?) and how is the method supposed to indicate the resulting status (ok or an error). E.g. is there a common exception they should be raising? Is there a common return value they should be returning (I see the return value is None, this is just an example of what "contract" means.)

Essentially a developer implementing the next child class should be able to look at this doc comment and know exactly how does the implementation have to behave in order to fulfill the API.

Reply:
I can update the comment.  We can make the method raise an exception if the benchmark fails since we are at the point where the overall logic works in Salvo.  We still see some NightHawk 'failures' where Envoy generates a `SIGTERM` at the end of a test and did not want to chase those issues while working on Salvo.

Reply:
Thank you, the documentation in the latest commit answered all the questions.

Resolving this comment thread, but leaving a separate comment about the return value.